### ISSUE TITLE
[BUG] Unplugging USB keyboard hides errors and doesn't always remove from DeviceRegistry

### ISSUE THREAD

**Issue by RedBearAK** [joshgoebel/keyszer#82]
See TODO comment on #83 to see what still needs to be done here.

---

A followup to issue #77 (closed). 

So, the simple and obvious explanation for the USB keyboards not being picked up automatically was that I didn't have "-w" or "--watch" in the command that runs `keyszer`. I must be misremembering but I thought you said at some point that "--watch" was no longer going to be necessary with `keyszer`, which was probably why I removed it. Seems at the moment it's still necessary to connect and disconnect devices on the fly. 

(Edit: I think it's "--quiet" that was actually removed and replaced with "--verbose". I probably mistakenly removed "--watch" around the same time.) 

But, even with that CLI option added, there are some exceptions showing up in the terminal when disconnecting a USB keyboard: 

```python
(II) in F (release)
(DD) on_key F release
(OO) release F
(DD) ###  device_change | registry, events: (<keyszer.devices.DeviceRegistry object at 0x7fdeb4c64a90>, [Event(wd=1, mask=4, cookie=0, name='event16'), Event(wd=1, mask=512, cookie=0, name='event16'), Event(wd=1, mask=4, cookie=0, name='event17'), Event(wd=1, mask=512, cookie=0, name='event17'), Event(wd=1, mask=4, cookie=0, name='event18'), Event(wd=1, mask=512, cookie=0, name='event18')])  ###

(II) in BACKSPACE (press)
(DD) on_key BACKSPACE press
(OO) press BACKSPACE
 
(II) in BACKSPACE (release)
(DD) on_key BACKSPACE release
(OO) release BACKSPACE
Traceback (most recent call last):
  File "/home/kris/Downloads/josh-keyszer/keyszer/.venv/lib/python3.10/site-packages/evdev/device.py", line 125, in __init__
    fd = os.open(dev, os.O_RDWR | os.O_NONBLOCK)
FileNotFoundError: [Errno 2] No such file or directory: '/dev/input/event16'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kris/Downloads/josh-keyszer/keyszer/src/keyszer/input.py", line 156, in device_change
    device = InputDevice(filename)
  File "/home/kris/Downloads/josh-keyszer/keyszer/.venv/lib/python3.10/site-packages/evdev/device.py", line 127, in __init__
    fd = os.open(dev, os.O_RDONLY | os.O_NONBLOCK)
FileNotFoundError: [Errno 2] No such file or directory: '/dev/input/event16'
```

These exceptions don't seem to actually interfere with the continued functioning of `keyszer` after they appear. So they are just incidental, I guess? Non-terminal exceptions? Something like that. 

The errors appear a few seconds after unplugging the keyboard. 

The same keyboard can be plugged back in and gets "grabbed" like it normally would, and the remapping seems to work like it normally would also.

**Comment by joshgoebel** [comment]
You could have a look at the line mentioned in the traceback in `input.py` and try to add a try/except block to wrap it and see if that makes the errors go away.  We already do this at the bottom, but not for the creation of the InputDevice itself.

**Comment by RedBearAK** [comment]
> You could have a look at the line mentioned in the traceback in input.py and try to add a try/except block to wrap it and see if that makes the errors go away. We already do this at the bottom, but not for the creation of the InputDevice itself.

I have no doubt that may make the error disappear, but I would feel uncomfortable calling that a solution to what's happening. That seems like just telling a cookie thief not to make a big fuss when they don't find anything in the cookie jar. 😂 

If I understand this error, line 156 in input is asking evdev to return the full device string for a device that has already registered a disconnect event. I would think this shouldn't be attempted at all with the device path of something that doesn't exist. I'll have to stare at the code with a blank look on my face for a while and see if I can't suss out a way to validate the existence of the `filename` before attempting to use it.

**Comment by joshgoebel** [comment]
> and see if I can't suss out a way to validate the existence of the filename before attempting to use it.

You can't in this scenario - that's the problem... this is a timing/multitasking issue... even if you added a check:

```rb
if File.exist?(device_name):
  # ok now it's safe to read
```

**It's not safe to read.**  The file could be removed right AFTER you check that it exists - because our process might have gotten suspended (briefly) by the OS while other things happen... the best way to check if it exists is just to call `InputDevice` and try to open it...  and the properly deal with the error case.

**Comment by RedBearAK** [comment]
> You can't in this scenario - that's the problem... this is a timing/multitasking issue... even if you added a check:

> The file could be removed by the OS right AFTER you check that it exists - because our process might have gotten suspended in-between those two lines... the best way to check if it exists is just to call InputDevice and try to open it... and the properly deal with the error case.

I kind of see and don't see at the same time. Make sense that there should really be a try/except wrapper in either case, but there is output in the log a few seconds before the error (part of what I pasted) that seems to indicate that some part of the code already knows the device was disconnected. Shouldn't that information about the disconnection event just be made available within the function that attempts to access the device? 

Or, did the action that ultimately generates the error in the log actually start happening before the log shows the disconnection event, because something is happening asynchronously?

**Comment by RedBearAK** [comment]
This seems to work without causing any problems I've identified immediately. 

```python
        filename = f"/dev/input/{event.name}"
        try:
            device = InputDevice(filename)
        except FileNotFoundError:
            continue
```

**Comment by joshgoebel** [comment]
I'd probably add a small comment linking to this issue... but yes, that's exactly what I imagined.  The continue being important so we move on to the next event in the loop to process.

**Comment by joshgoebel** [comment]
> but there is output in the log a few seconds before the error 

Is it really a few seconds?  If you read the notify handler code you'll see we purposely wait 0.5 seconds for notify events to "settle" and give the filesystem time to be updated FULLY... this often helps when adding new hardware, but likely hurts us when removing hardware.

Hmmm... actually this could leave dead devices hanging around inside our registry, but I'm trying to think now of if that would actually be super harmful and how we might avoid that given the fact it's hard to get the actual device object at that point (after it's node is gone)...

Hmmmm.

**Comment by RedBearAK** [comment]
During testing I watched the log in the terminal (with the new flushing active, so it tends to be highly responsive now) and more than once it took at least 2 seconds for the error to appear after disconnecting the keyboard. It was like, "Oh, I guess everything is fine--whoops, what's this?"

**Comment by RedBearAK** [comment]
Submitted PR #83 to fix. The log error, at least.

**Comment by RedBearAK** [comment]
@joshgoebel 

I forgot about this and thought it was merged a while ago. Is there a problem with the solution?

**Comment by joshgoebel** [comment]
I was holding it cause it wasn't a full fix, but I merged.  Will leave this open to track the larger issue.

**Comment by RedBearAK** [comment]
Got what appeared to be a race condition after unplugging a USB keyboard. I heard the fan turn on, which is pretty unusual on this laptop. Checked `btop` and keyszer was maxing out a couple of cores, using about 12% CPU constantly (or maybe just the process moving from one core to another, but maxing out one at a time). I had disconnected a USB keyboard about 30 seconds to a minute before I noticed the fan noise increasing. 

I bailed and restarted keyszer and it's back to between zero and 0.5% CPU, depending on keyboard usage. 

Definitely something odd going on when disconnecting a USB keyboard. I'm trying to finish another function right now, so I don't have time to look deeper into it.
